src/db_options.rs

```diff
@@ -4009,6 +4009,78 @@ impl CompactOptions {
     }
 }
 
+pub struct WaitForCompactOptions {
+    pub(crate) inner: *mut ffi::rocksdb_wait_for_compact_options_t,
+}
+
+impl Default for WaitForCompactOptions {
+    fn default() -> Self {
+        let opts = unsafe { ffi::rocksdb_wait_for_compact_options_create() };
+        assert!(
+            !opts.is_null(),
+            "Could not create RocksDB Wait For Compact Options"
+        );
+
+        Self { inner: opts }
+    }
+}
+
+impl Drop for WaitForCompactOptions {
+    fn drop(&mut self) {
+        unsafe {
+            ffi::rocksdb_wait_for_compact_options_destroy(self.inner);
+        }
+    }
+}
+
+impl WaitForCompactOptions {
+    /// If true, abort waiting if background jobs are paused. If false,
+    /// ContinueBackgroundWork() must be called to resume the background jobs.
+    /// Otherwise, jobs that were queued, but not scheduled yet may never finish
+    /// and WaitForCompact() may wait indefinitely (if timeout is set, it will
+    /// abort after the timeout).
+    ///
+    /// Default: false
+    pub fn set_abort_on_pause(&mut self, v: bool) {
+        unsafe {
+            ffi::rocksdb_wait_for_compact_options_set_abort_on_pause(self.inner, c_uchar::from(v));
+        }
+    }
+
+    /// If true, flush all column families before starting to wait.
+    ///
+    /// Default: false
+    pub fn set_flush(&mut self, v: bool) {
+        unsafe {
+            ffi::rocksdb_wait_for_compact_options_set_flush(self.inner, c_uchar::from(v));
+        }
+    }
+
+    /// If true, close db after waiting is done. By the time Close() is
+    /// called here, there should be no background jobs in progress and no new
+    /// background jobs should be added. DB may not have been closed if Close()
+    /// returned Aborted status due to unreleased snapshots in the system. See
+    /// comments in DB::Close() for details.
+    ///
+    /// Default: false
+    pub fn set_close_db(&mut self, v: bool) {
```

Comment:
I think method shouldn't be exposed, or alternatively should be marked unsafe? It allows the underlying database to be closed without dropping the DB object, and RocksDB is pretty light on guarantees about continuing to use a DB after it was closed.

https://github.com/facebook/rocksdb/blob/c96d9a0fbbfd2215e7ae88af8a13b94109a349b5/include/rocksdb/db.h#L345

> Regardless of the return status, the DB must be freed.

From the wiki:

> Q: Is it safe to close RocksDB while another thread is issuing read, write or manual compaction requests?
> 
> A: No. The users of RocksDB need to make sure all functions have finished before they close RocksDB. You can speed up the waiting by calling CancelAllBackgroundWork().

Looking at the implementation, there *is* a `closed_` flag in RocksDB but it is only checked in Close and in the destructor

https://github.com/facebook/rocksdb/blob/c96d9a0fbbfd2215e7ae88af8a13b94109a349b5/db/db_impl/db_impl.cc#L5136

Reply:
Another safe alternative would be remove this and add a `pub fn wait_for_compact_and_close(self, opts: &WaitForCompactOptions) -> Result<(), Error>` method that consumes the DB instance

Reply:
hmmm, ok for the time being, I'm going to remove set_close_db as I don't need it. We can revist if someone needs this option